processing/src/main/java/io/druid/query/groupby/having/DimFilterHavingSpec.java

```diff
@@ -25,26 +25,29 @@
 import io.druid.data.input.Row;
 import io.druid.query.filter.DimFilter;
 import io.druid.query.filter.ValueMatcher;
-import io.druid.query.groupby.RowBasedValueMatcherFactory;
+import io.druid.query.groupby.RowBasedColumnSelectorFactory;
+import io.druid.segment.column.ValueType;
 
 import java.nio.ByteBuffer;
+import java.util.Map;
 
-public class DimFilterHavingSpec implements HavingSpec
+public class DimFilterHavingSpec extends BaseHavingSpec
 {
   private static final byte CACHE_KEY = (byte) 0x9;
 
   private final DimFilter dimFilter;
-  private final RowBasedValueMatcherFactory valueMatcherFactory;
-  private final ValueMatcher valueMatcher;
+  private final ThreadLocal<Row> rowSupplier;
+
+  private final Object valueMatcherMonitor = new Object();
+  private volatile ValueMatcher valueMatcher;
```

Comment:
Though loop example is misleading, because even without loops, a thread is not required to read the "most recent" value from volatile, because there is no global order of volatile reads/writes in the program, so the "most recent" is not defined. Only if somebody write value `x` to a volatile field in a thread `A` and somebody else read that `x` from that field in a thread `B`, it forms a happens-before edge and allows to build partial orders. But thread `B` is not *required* to *ever* read that `x`.